ethpm_cli/manifest.py

```diff
@@ -81,45 +86,63 @@ def gen_validate_manifest() -> Optional[Callable[..., Manifest]]:
     return None
 
 
-def gen_contract_types(
-    solc_output: Dict[str, Any]
-) -> Iterable[Callable[..., Manifest]]:
-    contract_types = get_contract_types(solc_output)
-    pretty = "\n".join(contract_types)
+def gen_contract_types_and_sources(
+    solc_output: Dict[str, Any], contracts_dir: Path
+) -> Tuple[Callable[..., Manifest], ...]:
+    # todo: option to include additional sources not associated with included contract types
+    ctypes_and_sources = get_contract_types_and_sources(solc_output)
+    all_contract_types = [ct[0] for ct in ctypes_and_sources]
+    pretty = "".join(format_contract_types_and_sources_for_display(ctypes_and_sources))
     flag = parse_bool_flag(
         "\n"
-        f"{len(contract_types)} contract types available.\n\n"
-        f"{pretty}. \n\n"
-        "Would you like to include all available contract types?"
+        f"{len(all_contract_types)} contract types available.\n\n"
+        f"{pretty}\n"
+        "Would you like to automatically include all available contract types and their sources?"
     )
-    if not flag:
-        raise Exception("Custom contract types are not supported yet.")
-    return build_contract_types(contract_types, solc_output)
 
+    if flag:
+        target_contract_types = all_contract_types
+    else:
+        while True:
+            raw_included_ctypes = input(
+                "Please list the contract types you would like to include, separated by commas: "
+            )
+            target_contract_types = [
+                ct.strip(" ") for ct in raw_included_ctypes.split(",")
+            ]
+            if any(ct not in all_contract_types for ct in target_contract_types):
+                cli_logger.info(f"Invalid contract type selected. Please try again.")
+            else:
+                break
 
-def gen_sources(
-    solc_output: Dict[str, Any], contracts_dir: Path
-) -> Iterable[Callable[..., Manifest]]:
-    available_sources = [
-        str(src.relative_to(contracts_dir)) for src in contracts_dir.glob("**/*.sol")
-    ]
-    contract_types = get_contract_types(solc_output)
-    pretty = "\n".join(sorted(available_sources))
-    flag = parse_bool_flag(
-        "\n"
-        f"{len(available_sources)} sources available.\n\n"
-        f"{pretty}. \n\n"
-        "Would you like to include all available sources?"
-    )
-    if not flag:
-        raise Exception("sorry, we dont support specific source selection yet.")
     inline_source_flag = parse_bool_flag(
-        "Would you like to automatically inline all sources?"
+        "Would you like to inline source files? If not, sources will "
+        "be automatically pinned to IPFS."
+    )
+    target_sources = set(
+        flatten([ct[1] for ct in ctypes_and_sources if ct[0] in target_contract_types])
```

Comment:
Maybe rather than indexing into `ct` you could do:

```
flatten(source_path for ctype, source_path in ctypes_and_sources if ctype in target_contract_types)
```